src/connection.ts

```diff
@@ -647,30 +576,21 @@ export class StableWSConnection<
    * _setupPromise - sets up the this.connectOpen promise
    */
   _setupConnectionPromise = () => {
-    const that = this;
     this.isResolved = false;
     /** a promise that is resolved once ws.open is called */
-    this.connectionOpen = new Promise<WebSocket.MessageEvent>(function (resolve, reject) {
-      that.resolvePromise = resolve;
-      that.rejectPromise = reject;
-    }).then(
-      (e) => {
-        if (e.data && typeof e.data === 'string') {
-          const data = JSON.parse(e.data) as ConnectionOpen<ChannelType, CommandType, UserType> & {
-            error?: unknown;
-          };
-          if (data && data.error != null) {
-            throw new Error(JSON.stringify(data.error));
-          }
-          return data;
-        } else {
-          return undefined;
-        }
-      },
-      (error) => {
-        throw error;
-      },
-    );
+    this.connectionOpen = new Promise<WebSocket.MessageEvent>((resolve, reject) => {
+      this.resolvePromise = resolve;
+      this.rejectPromise = reject;
+    }).then((e) => {
+      if (!e.data || typeof e.data !== 'string') return;
+
+      const data = JSON.parse(e.data) as ConnectionOpen<ChannelType, CommandType, UserType> & { error?: unknown };
+      if (data?.error) {
+        throw new Error(JSON.stringify(data.error));
+      }
```

Comment:
I think we can also get rid of this, since we already have this check in `onMessage` function ... and thats the only place where resolvePromise gets called.

```tsx
  onmessage = (wsID: number, event: WebSocket.MessageEvent) => {
    if (this.wsID !== wsID) return;

    const data = typeof event.data === 'string' ? JSON.parse(event.data) : null;

    // we wait till the first message before we consider the connection open..
    // the reason for this is that auth errors and similar errors trigger a ws.onopen and immediately
    // after that a ws.onclose..
    if (!this.isResolved && data) {
      this.isResolved = true;
      if (data.error != null) { // <-------- THIS ONE
        this.rejectPromise?.(this._errorFromWSEvent(data, false));
        return;
      } else {
        this.resolvePromise?.(event);
        this._setHealth(true);
      }
    }

    // trigger the event..
    this.lastEvent = new Date();
    this.logger('info', 'connection:onmessage() - onmessage callback', {
      tags: ['connection'],
      event,
      wsID,
    });

    if (data && data.type === 'health.check') {
      this.scheduleNextPing();
    }

    this.messageCallback(event);
    this.scheduleConnectionCheck();
  };
```

Reply:
Check for error I mean!!

Reply:
It also opens a possibility of getting rid of this `then` block.

Reply:
Or even `_setupConnectionPromise`, not 100% sure though ... just wild idea

Reply:
good stuff, removed duplication here https://github.com/GetStream/stream-chat-js/pull/822/commits/f672fdc8fa9a82199d7b9650eabad4143db291cd